### ISSUE TITLE
Compile errors

### ISSUE THREAD

**Issue by zciendor** [mwarning/KadNode#85]
I can't compile KadNode on 32-bit ARM, especially a [zHIVErbox](https://github.com/zhiverbox/armbian-userpatches). `mbedtls-2.12.0` was compiled and installed without issues first.

```
$ git checkout -b v2.2.3
$ make

cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/main.o src/main.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/searches.o src/searches.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/kad.o src/kad.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/log.o src/log.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/conf.o src/conf.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/net.o src/net.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/utils.o src/utils.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/announces.o src/announces.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/peerfile.o src/peerfile.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/unix.o src/unix.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-lpd.o src/ext-lpd.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-bob.o src/ext-bob.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-cmd.o src/ext-cmd.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-dns.o src/ext-dns.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-nss.o src/ext-nss.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-tls-client.o src/ext-tls-client.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -c -o build/ext-tls-server.o src/ext-tls-server.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -fPIC -c -o build/ext-libnss.o src/ext-libnss.c
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -fPIC -shared -Wl,-soname,libnss_kadnode.so.2 -o build/libnss_kadnode-2.0.so build/ext-libnss.o
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto build/main.o build/searches.o build/kad.o build/log.o build/conf.o build/net.o build/utils.o build/announces.o build/peerfile.o build/unix.o build/ext-lpd.o build/ext-bob.o build/ext-cmd.o build/ext-dns.o build/ext-nss.o build/ext-tls-client.o build/ext-tls-server.o -o build/kadnode
build/ext-bob.o: In function `get_pkey_base32hex':
/opt/src/KadNode/src/ext-bob.c:321: undefined reference to `mbedtls_mpi_write_binary'
build/ext-bob.o: In function `bob_send_challenge':
/opt/src/KadNode/src/ext-bob.c:209: undefined reference to `mbedtls_mpi_write_binary'
build/ext-bob.o: In function `mbedtls_ecp_decompress':
/opt/src/KadNode/src/ext-bob.c:87: undefined reference to `mbedtls_mpi_size'
/opt/src/KadNode/src/ext-bob.c:104: undefined reference to `mbedtls_mpi_init'
/opt/src/KadNode/src/ext-bob.c:105: undefined reference to `mbedtls_mpi_init'
/opt/src/KadNode/src/ext-bob.c:106: undefined reference to `mbedtls_mpi_init'
/opt/src/KadNode/src/ext-bob.c:109: undefined reference to `mbedtls_mpi_read_binary'
/opt/src/KadNode/src/ext-bob.c:112: undefined reference to `mbedtls_mpi_mul_mpi'
/opt/src/KadNode/src/ext-bob.c:117: undefined reference to `mbedtls_mpi_sub_int'
/opt/src/KadNode/src/ext-bob.c:123: undefined reference to `mbedtls_mpi_mul_mpi'
/opt/src/KadNode/src/ext-bob.c:126: undefined reference to `mbedtls_mpi_add_mpi'
/opt/src/KadNode/src/ext-bob.c:132: undefined reference to `mbedtls_mpi_add_int'
/opt/src/KadNode/src/ext-bob.c:135: undefined reference to `mbedtls_mpi_shift_r'
/opt/src/KadNode/src/ext-bob.c:138: undefined reference to `mbedtls_mpi_exp_mod'
/opt/src/KadNode/src/ext-bob.c:141: undefined reference to `mbedtls_mpi_get_bit'
/opt/src/KadNode/src/ext-bob.c:147: undefined reference to `mbedtls_mpi_write_binary'
/opt/src/KadNode/src/ext-bob.c:150: undefined reference to `mbedtls_mpi_free'
/opt/src/KadNode/src/ext-bob.c:151: undefined reference to `mbedtls_mpi_free'
/opt/src/KadNode/src/ext-bob.c:152: undefined reference to `mbedtls_mpi_free'
/opt/src/KadNode/src/ext-bob.c:119: undefined reference to `mbedtls_mpi_add_mpi'
/opt/src/KadNode/src/ext-bob.c:143: undefined reference to `mbedtls_mpi_sub_mpi'
build/ext-bob.o: In function `bob_trigger_auth':
/opt/src/KadNode/src/ext-bob.c:266: undefined reference to `mbedtls_ecp_point_read_binary'
build/ext-bob.o: In function `bob_create_key':
/opt/src/KadNode/src/ext-bob.c:336: undefined reference to `mbedtls_pk_init'
/opt/src/KadNode/src/ext-bob.c:337: undefined reference to `mbedtls_ctr_drbg_init'
/opt/src/KadNode/src/ext-bob.c:338: undefined reference to `mbedtls_entropy_init'
/opt/src/KadNode/src/ext-bob.c:341: undefined reference to `mbedtls_ctr_drbg_seed'
/opt/src/KadNode/src/ext-bob.c:349: undefined reference to `mbedtls_pk_info_from_type'
/opt/src/KadNode/src/ext-bob.c:349: undefined reference to `mbedtls_pk_setup'
/opt/src/KadNode/src/ext-bob.c:357: undefined reference to `mbedtls_ecp_gen_key'
/opt/src/KadNode/src/ext-bob.c:362: undefined reference to `mbedtls_mpi_get_bit'
build/ext-bob.o: In function `write_pem':
/opt/src/KadNode/src/ext-bob.c:289: undefined reference to `mbedtls_pk_write_key_pem'
build/ext-bob.o: In function `bob_create_key':
/opt/src/KadNode/src/ext-bob.c:372: undefined reference to `mbedtls_entropy_func'
/opt/src/KadNode/src/ext-bob.c:372: undefined reference to `mbedtls_ctr_drbg_random'
build/ext-bob.o: In function `bob_load_key':
/opt/src/KadNode/src/ext-bob.c:381: undefined reference to `mbedtls_pk_init'
/opt/src/KadNode/src/ext-bob.c:383: undefined reference to `mbedtls_pk_parse_keyfile'
/opt/src/KadNode/src/ext-bob.c:384: undefined reference to `mbedtls_pk_free'
/opt/src/KadNode/src/ext-bob.c:385: undefined reference to `mbedtls_strerror'
/opt/src/KadNode/src/ext-bob.c:391: undefined reference to `mbedtls_ecp_curve_info_from_grp_id'
build/ext-bob.o: In function `bob_verify_challenge':
/opt/src/KadNode/src/ext-bob.c:456: undefined reference to `mbedtls_ecdsa_read_signature'
build/ext-bob.o: In function `bob_find_key':
/opt/src/KadNode/src/ext-bob.c:471: undefined reference to `mbedtls_mpi_write_binary'
build/ext-bob.o: In function `bob_encrypt_challenge':
/opt/src/KadNode/src/ext-bob.c:498: undefined reference to `mbedtls_ecdsa_write_signature'
/opt/src/KadNode/src/ext-bob.c:511: undefined reference to `mbedtls_ctr_drbg_random'
build/ext-bob.o: In function `bob_setup':
/opt/src/KadNode/src/ext-bob.c:571: undefined reference to `mbedtls_ctr_drbg_init'
/opt/src/KadNode/src/ext-bob.c:572: undefined reference to `mbedtls_entropy_init'
/opt/src/KadNode/src/ext-bob.c:577: undefined reference to `mbedtls_pk_info_from_type'
/opt/src/KadNode/src/ext-bob.c:577: undefined reference to `mbedtls_pk_setup'
/opt/src/KadNode/src/ext-bob.c:578: undefined reference to `mbedtls_ecp_group_load'
build/ext-bob.o: In function `bob_free':
/opt/src/KadNode/src/ext-bob.c:598: undefined reference to `mbedtls_ctr_drbg_free'
/opt/src/KadNode/src/ext-bob.c:599: undefined reference to `mbedtls_entropy_free'
/opt/src/KadNode/src/ext-bob.c:605: undefined reference to `mbedtls_pk_free'
build/ext-tls-client.o: In function `tls_conf_verify':
/opt/src/KadNode/src/ext-tls-client.c:287: undefined reference to `mbedtls_x509_crt_info'
/opt/src/KadNode/src/ext-tls-client.c:290: undefined reference to `mbedtls_x509_crt_verify_info'
build/ext-tls-client.o: In function `tls_connect_init':
/opt/src/KadNode/src/ext-tls-client.c:57: undefined reference to `mbedtls_ssl_set_bio'
/opt/src/KadNode/src/ext-tls-client.c:59: undefined reference to `mbedtls_ssl_set_hostname'
/opt/src/KadNode/src/ext-tls-client.c:70: undefined reference to `mbedtls_net_set_nonblock'
/opt/src/KadNode/src/ext-tls-client.c:80: undefined reference to `mbedtls_net_free'
/opt/src/KadNode/src/ext-tls-client.c:81: undefined reference to `mbedtls_net_init'
build/ext-tls-client.o: In function `tls_client_trigger_auth':
/opt/src/KadNode/src/ext-tls-client.c:273: undefined reference to `mbedtls_net_recv'
/opt/src/KadNode/src/ext-tls-client.c:273: undefined reference to `mbedtls_net_send'
build/ext-tls-client.o: In function `auth_end':
/opt/src/KadNode/src/ext-tls-client.c:111: undefined reference to `mbedtls_ssl_close_notify'
/opt/src/KadNode/src/ext-tls-client.c:116: undefined reference to `mbedtls_net_free'
/opt/src/KadNode/src/ext-tls-client.c:117: undefined reference to `mbedtls_ssl_session_reset'
/opt/src/KadNode/src/ext-tls-client.c:120: undefined reference to `mbedtls_net_init'
build/ext-tls-client.o: In function `tls_handle':
/opt/src/KadNode/src/ext-tls-client.c:159: undefined reference to `mbedtls_ssl_handshake'
/opt/src/KadNode/src/ext-tls-client.c:179: undefined reference to `mbedtls_ssl_get_version'
/opt/src/KadNode/src/ext-tls-client.c:179: undefined reference to `mbedtls_ssl_get_ciphersuite'
/opt/src/KadNode/src/ext-tls-client.c:179: undefined reference to `mbedtls_ssl_get_max_frag_len'
/opt/src/KadNode/src/ext-tls-client.c:185: undefined reference to `mbedtls_ssl_get_verify_result'
/opt/src/KadNode/src/ext-tls-client.c:191: undefined reference to `mbedtls_x509_crt_verify_info'
/opt/src/KadNode/src/ext-tls-client.c:195: undefined reference to `mbedtls_ssl_get_peer_cert'
/opt/src/KadNode/src/ext-tls-client.c:196: undefined reference to `mbedtls_ssl_get_peer_cert'
/opt/src/KadNode/src/ext-tls-client.c:196: undefined reference to `mbedtls_x509_crt_info'
build/ext-tls-client.o: In function `tls_client_get_id':
/opt/src/KadNode/src/ext-tls-client.c:212: undefined reference to `mbedtls_sha256_init'
/opt/src/KadNode/src/ext-tls-client.c:218: undefined reference to `mbedtls_sha256_update'
/opt/src/KadNode/src/ext-tls-client.c:219: undefined reference to `mbedtls_sha256_finish'
build/ext-tls-client.o: In function `tls_client_add_ca':
/opt/src/KadNode/src/ext-tls-client.c:308: undefined reference to `mbedtls_x509_crt_init'
/opt/src/KadNode/src/ext-tls-client.c:312: undefined reference to `mbedtls_x509_crt_parse_file'
/opt/src/KadNode/src/ext-tls-client.c:312: undefined reference to `mbedtls_x509_crt_parse_path'
/opt/src/KadNode/src/ext-tls-client.c:314: undefined reference to `mbedtls_strerror'
build/ext-tls-client.o: In function `tls_client_setup':
/opt/src/KadNode/src/ext-tls-client.c:342: undefined reference to `mbedtls_ctr_drbg_init'
/opt/src/KadNode/src/ext-tls-client.c:343: undefined reference to `mbedtls_entropy_init'
/opt/src/KadNode/src/ext-tls-client.c:345: undefined reference to `mbedtls_ctr_drbg_seed'
/opt/src/KadNode/src/ext-tls-client.c:352: undefined reference to `mbedtls_ssl_init'
/opt/src/KadNode/src/ext-tls-client.c:353: undefined reference to `mbedtls_net_init'
/opt/src/KadNode/src/ext-tls-client.c:352: undefined reference to `mbedtls_ssl_init'
/opt/src/KadNode/src/ext-tls-client.c:353: undefined reference to `mbedtls_net_init'
/opt/src/KadNode/src/ext-tls-client.c:356: undefined reference to `mbedtls_ssl_config_init'
/opt/src/KadNode/src/ext-tls-client.c:359: undefined reference to `mbedtls_ssl_config_defaults'
/opt/src/KadNode/src/ext-tls-client.c:368: undefined reference to `mbedtls_ssl_conf_verify'
/opt/src/KadNode/src/ext-tls-client.c:371: undefined reference to `mbedtls_ssl_conf_rng'
/opt/src/KadNode/src/ext-tls-client.c:372: undefined reference to `mbedtls_ssl_conf_read_timeout'
/opt/src/KadNode/src/ext-tls-client.c:373: undefined reference to `mbedtls_ssl_conf_ca_chain'
/opt/src/KadNode/src/ext-tls-client.c:377: undefined reference to `mbedtls_ssl_setup'
/opt/src/KadNode/src/ext-tls-client.c:377: undefined reference to `mbedtls_ssl_setup'
/opt/src/KadNode/src/ext-tls-client.c:379: undefined reference to `mbedtls_entropy_func'
/opt/src/KadNode/src/ext-tls-client.c:379: undefined reference to `mbedtls_ctr_drbg_random'
build/ext-tls-client.o: In function `tls_client_free':
/opt/src/KadNode/src/ext-tls-client.c:391: undefined reference to `mbedtls_ssl_free'
/opt/src/KadNode/src/ext-tls-client.c:392: undefined reference to `mbedtls_net_free'
/opt/src/KadNode/src/ext-tls-client.c:391: undefined reference to `mbedtls_ssl_free'
/opt/src/KadNode/src/ext-tls-client.c:392: undefined reference to `mbedtls_net_free'
/opt/src/KadNode/src/ext-tls-client.c:395: undefined reference to `mbedtls_x509_crt_free'
/opt/src/KadNode/src/ext-tls-client.c:396: undefined reference to `mbedtls_ssl_config_free'
/opt/src/KadNode/src/ext-tls-client.c:397: undefined reference to `mbedtls_entropy_free'
/opt/src/KadNode/src/ext-tls-client.c:398: undefined reference to `mbedtls_ctr_drbg_free'
build/ext-tls-server.o: In function `tls_server_handler':
/opt/src/KadNode/src/ext-tls-server.c:158: undefined reference to `mbedtls_net_accept'
/opt/src/KadNode/src/ext-tls-server.c:166: undefined reference to `mbedtls_net_set_nonblock'
/opt/src/KadNode/src/ext-tls-server.c:172: undefined reference to `mbedtls_ssl_conf_read_timeout'
/opt/src/KadNode/src/ext-tls-server.c:174: undefined reference to `mbedtls_ssl_set_bio'
/opt/src/KadNode/src/ext-tls-server.c:177: undefined reference to `mbedtls_net_recv'
/opt/src/KadNode/src/ext-tls-server.c:177: undefined reference to `mbedtls_net_send'
build/ext-tls-server.o: In function `sni_callback':
/opt/src/KadNode/src/ext-tls-server.c:227: undefined reference to `mbedtls_ssl_set_hs_authmode'
build/ext-tls-server.o: In function `end_client_connection':
/opt/src/KadNode/src/ext-tls-server.c:70: undefined reference to `mbedtls_ssl_close_notify'
/opt/src/KadNode/src/ext-tls-server.c:73: undefined reference to `mbedtls_net_free'
/opt/src/KadNode/src/ext-tls-server.c:74: undefined reference to `mbedtls_ssl_session_reset'
/opt/src/KadNode/src/ext-tls-server.c:83: undefined reference to `mbedtls_strerror'
build/ext-tls-server.o: In function `tls_client_handler':
/opt/src/KadNode/src/ext-tls-server.c:103: undefined reference to `mbedtls_ssl_handshake'
/opt/src/KadNode/src/ext-tls-server.c:115: undefined reference to `mbedtls_ssl_get_verify_result'
/opt/src/KadNode/src/ext-tls-server.c:116: undefined reference to `mbedtls_x509_crt_verify_info'
/opt/src/KadNode/src/ext-tls-server.c:123: undefined reference to `mbedtls_ssl_get_version'
/opt/src/KadNode/src/ext-tls-server.c:123: undefined reference to `mbedtls_ssl_get_ciphersuite'
/opt/src/KadNode/src/ext-tls-server.c:126: undefined reference to `mbedtls_ssl_get_record_expansion'
build/ext-tls-server.o: In function `tls_server_add_sni':
/opt/src/KadNode/src/ext-tls-server.c:249: undefined reference to `mbedtls_x509_crt_init'
/opt/src/KadNode/src/ext-tls-server.c:250: undefined reference to `mbedtls_pk_init'
/opt/src/KadNode/src/ext-tls-server.c:252: undefined reference to `mbedtls_x509_crt_parse_file'
/opt/src/KadNode/src/ext-tls-server.c:253: undefined reference to `mbedtls_strerror'
/opt/src/KadNode/src/ext-tls-server.c:258: undefined reference to `mbedtls_pk_parse_keyfile'
/opt/src/KadNode/src/ext-tls-server.c:259: undefined reference to `mbedtls_strerror'
build/ext-tls-server.o: In function `get_common_name':
/opt/src/KadNode/src/ext-tls-server.c:190: undefined reference to `mbedtls_oid_get_attr_short_name'
build/ext-tls-server.o: In function `tls_server_add_sni':
/opt/src/KadNode/src/ext-tls-server.c:292: undefined reference to `mbedtls_x509_crt_info'
build/ext-tls-server.o: In function `tls_server_setup':
/opt/src/KadNode/src/ext-tls-server.c:334: undefined reference to `mbedtls_net_init'
/opt/src/KadNode/src/ext-tls-server.c:335: undefined reference to `mbedtls_net_init'
/opt/src/KadNode/src/ext-tls-server.c:336: undefined reference to `mbedtls_net_init'
/opt/src/KadNode/src/ext-tls-server.c:337: undefined reference to `mbedtls_net_init'
/opt/src/KadNode/src/ext-tls-server.c:339: undefined reference to `mbedtls_ssl_init'
/opt/src/KadNode/src/ext-tls-server.c:340: undefined reference to `mbedtls_ssl_config_init'
/opt/src/KadNode/src/ext-tls-server.c:341: undefined reference to `mbedtls_ctr_drbg_init'
/opt/src/KadNode/src/ext-tls-server.c:348: undefined reference to `mbedtls_entropy_init'
/opt/src/KadNode/src/ext-tls-server.c:349: undefined reference to `mbedtls_ctr_drbg_seed'
/opt/src/KadNode/src/ext-tls-server.c:359: undefined reference to `mbedtls_ssl_config_defaults'
/opt/src/KadNode/src/ext-tls-server.c:368: undefined reference to `mbedtls_ssl_conf_authmode'
/opt/src/KadNode/src/ext-tls-server.c:369: undefined reference to `mbedtls_ssl_conf_rng'
/opt/src/KadNode/src/ext-tls-server.c:372: undefined reference to `mbedtls_ssl_conf_sni'
/opt/src/KadNode/src/ext-tls-server.c:374: undefined reference to `mbedtls_ssl_setup'
/opt/src/KadNode/src/ext-tls-server.c:380: undefined reference to `mbedtls_net_set_nonblock'
/opt/src/KadNode/src/ext-tls-server.c:385: undefined reference to `mbedtls_net_set_nonblock'
/opt/src/KadNode/src/ext-tls-server.c:386: undefined reference to `mbedtls_entropy_func'
/opt/src/KadNode/src/ext-tls-server.c:386: undefined reference to `mbedtls_ctr_drbg_random'
build/ext-tls-server.o: In function `sni_callback':
/opt/src/KadNode/src/ext-tls-server.c:230: undefined reference to `mbedtls_ssl_set_hs_own_cert'
collect2: error: ld returned 1 exit status
Makefile:96: recipe for target 'kadnode' failed
make: *** [kadnode] Error 1
```

**Comment by mwarning** [comment]
It looks like the compiler does not find the shared library.
Where is libmbedtls.so located on your system?

**Comment by zciendor** [comment]
/usr/lib/arm-linux-gnueabihf/libmbedtls.so

**Comment by mwarning** [comment]
Please try `LDFLAGS=/usr/lib/arm-linux-gnueabihf make`

**Comment by zciendor** [comment]
```
$ LDFLAGS=/usr/lib/arm-linux-gnueabihf make
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS /usr/lib/arm-linux-gnueabihf -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -fPIC -c -o build/ext-libnss.o src/ext-libnss.c
cc: warning: /usr/lib/arm-linux-gnueabihf: linker input file unused because linking not done
cc -Os -Wall -Wwrite-strings -pedantic -std=gnu99 -I/usr/local/include  -DLPD -DBOB -DCMD -g -DDEBUG -DDNS -DNSS -DTLS /usr/lib/arm-linux-gnueabihf -L/usr/local/lib -lc -lmbedtls -lmbedx509 -lmbedcrypto -lmbedtls -lmbedx509 -lmbedcrypto -fPIC -shared -Wl,-soname,libnss_kadnode.so.2 -o build/libnss_kadnode-2.0.so build/ext-libnss.o
/usr/bin/ld: cannot find /usr/lib/arm-linux-gnueabihf: File format not recognized
collect2: error: ld returned 1 exit status
Makefile:85: recipe for target 'libnss_kadnode' failed
make: *** [libnss_kadnode] Error 1
```

**Comment by mwarning** [comment]
Ah, sorry. I meant `LDFLAGS=-L/usr/lib/arm-linux-gnueabihf make`

**Comment by zciendor** [comment]
I installed an older version of mbedtls via apt from the Debian repository and now it worked.

```
sudo apt install -f libmbedtls10
```

**Comment by mwarning** [comment]
Interesting, is the install path of libmbedtls10 a different path now?

**Comment by zciendor** [comment]
No. It just ends with `.10`
```
$ dpkg-query -L libmbedtls10
/.
/usr
/usr/lib
/usr/lib/arm-linux-gnueabihf
/usr/lib/arm-linux-gnueabihf/libmbedtls.so.2.4.2
/usr/share
/usr/share/doc
/usr/share/doc/libmbedtls10
/usr/share/doc/libmbedtls10/changelog.Debian.gz
/usr/share/doc/libmbedtls10/changelog.gz
/usr/share/doc/libmbedtls10/copyright
/usr/lib/arm-linux-gnueabihf/libmbedtls.so.10
```

**Comment by zciendor** [comment]
I also noticed, that when you compile `mbedtls` from the github sources just via running `make` and `make install` no library file is created. Seems like you have to call `SHARED=1 make lib` to cause a `.so` file to be generated. Anyway, with the latest `mbedtls-2.12.0` branch I ran into another compile error, so I decided to stick with the current one from Debian, which is `libmbedtls-2.4.2`.

**Comment by mwarning** [comment]
Maybe this fixes your problem with mbedtls-2.12.0: https://github.com/mwarning/KadNode/commit/0ea13be9c226bafe15035f838e87c337430a453b
